cdap-ui/app/cdap/components/PreviewData/DataView/Table.tsx

```diff
@@ -0,0 +1,211 @@
+import ThemeWrapper from 'components/ThemeWrapper';
+import VirtualScroll from 'components/VirtualScroll';
+import T from 'i18n-react';
+import classnames from 'classnames';
+
+const I18N_PREFIX = 'features.PreviewData.DataView.Table';
+
+export const messageTextStyle = {
+  fontSize: '1.3rem !important',
+  margin: 'unset',
+  padding: '10px 5px',
+};
+export const styles = (theme): StyleRules => ({
+  root: {
+    display: 'inline-block',
+    height: 'auto',
+    marginTop: theme.spacing(1),
+  },
+  tableContainer: {
+    width: 'fit-content',
+    minWidth: '100%',
+  },
+  row: {
+    height: 40,
+    '&.oddRow': {
+      backgroundColor: theme.palette.grey['600'],
+    },
+  },
+  messageText: messageTextStyle,
+  headerRow: {
+    backgroundColor: theme.palette.grey['300'],
+    fontWeight: 500,
+    color: theme.palette.common.white,
+    fontSize: 14,
+  },
+  cell: {
+    textAlign: 'left',
+    height: '40px',
+    lineHeight: '40px',
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+    padding: '0px 10px',
+  },
+
+  // TO DO: Currently the width is fixed. Future plan is to let users vary the column widths
+  tableCell: {
+    width: '120px',
+    borderLeft: `1px solid ${theme.palette.grey['500']}`,
+    flexGrow: 1,
+  },
+  indexCell: {
+    width: '50px',
+  },
+});
+
+interface IDataTableProps extends WithStyles<typeof styles> {
+  headers?: string[];
+  records?: any[];
+  isInput?: boolean;
+  previewStatus?: string;
+  isCondition?: boolean;
+}
+
+const DataTableView: React.FC<IDataTableProps> = ({
+  classes,
+  headers,
+  records,
+  isInput,
+  previewStatus,
+  isCondition,
+}) => {
+  const getStatusMsg = () => {
+    let msg;
+    const recordType = isInput ? 'Input' : 'Output';
+    if (isCondition) {
+      msg = T.translate(`${I18N_PREFIX}.previewNotSupported`);
+    } else if (previewStatus === PREVIEW_STATUS.RUNNING || previewStatus === PREVIEW_STATUS.INIT) {
+      // preview is still running but there's no data yet
+      msg = T.translate(`${I18N_PREFIX}.previewRunning`, { recordType });
+    } else {
+      // not running preview but there is no preview data
+      msg = T.translate(`${I18N_PREFIX}.noPreviewRunning`, { recordType });
+    }
+    return msg;
+  };
+
+  // Used to stringify any non-string field values and field names.
+  // TO DO: Might not need to do this for field names, need to test with nested schemas
+  const format = (field: any) => {
+    if (typeof field === 'object') {
+      return JSON.stringify(field);
+    }
+    return field;
+  };
+  const renderList = (visibleNodeCount: number, startNode: number) => {
+    return records.slice(startNode, startNode + visibleNodeCount).map((record, i) => {
+      return (
+        <React.Fragment>
+          <Grid
+            container
+            direction="row"
+            wrap="nowrap"
+            justify="space-evenly"
+            className={classnames(classes.row, { oddRow: (i + startNode + 1) % 2 })}
```

Comment:
can we define variable for i + startNode + 1 since it's being used in multiple places?

Reply:
done!